Make HOSTNAME matching more permissive

DennisDenuto opened the pull request:
Related to: 
https://github.com/vmware-tanzu/carvel-imgpkg/issues/234
https://github.com/vmware-tanzu/carvel-imgpkg/issues/235

`IMGPKG_REGISTRY_HOSTNAME` matching now allows:
- schemes (http/https)
- v1 and v2 in the path to be the equivalent to just passing the hostname
- glob patterns in the HOSTNAME (currently wildcards are not supported in ports and paths)

# Before merging

to avoid merge difficulty, this PR should be merged after https://github.com/vmware-tanzu/carvel-imgpkg/pull/220



TODO:

- [ ] Update docs to reflect new behavior of `IMGPKG_REGISTRY_HOSTNAME`

cppforlife commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -49,7 +50,12 @@ func (k *envKeychain) Resolve(target regauthn.Resource) (regauthn.Authenticator,
 
 	for _, info := range infos {
-		if info.Hostname == target.RegistryStr() {
+		registryURLMatches, err := credentialprovider.URLsMatchStr(info.Hostname, target.RegistryStr())
```
might be good to document for users (lightly) how URLsMatchStr behaves.

DennisDenuto replied:
agreed. I had a TODO to document the new behavior of `IMGPKG_REGISTRY_HOSTNAME` 

I was thinking carvel.dev doc website. did you have something else in mind? (--help?)

cppforlife replied:
nah, just regular docs on carvel.dev

cppforlife commented on pkg/imgpkg/registry/keychain_test.go:
```diff
@@ -169,6 +170,58 @@ func TestAuthProvidedViaEnvVars(t *testing.T) {
+		"localhost:9999/v2/",
+		"*:9999/v2/",
+		"local*:9999/v2/",
+		"localhost:9999/imgpkg_test",
```
do we allow globs within path?

DennisDenuto replied:
No thats a limitation with the package (k8s-pkg-credentialprovider) we are using

https://github.com/vmware-tanzu/carvel-imgpkg/blob/3a779ca0886ab989a22c6928a1eb58cd36dc0bbd/vendor/github.com/vdemeester/k8s-pkg-credentialprovider/keyring.go#L203

cppforlife replied:
ah interesting, i guess thats ok for now. could always expand later. what happens if somebody does throw in a star into a path? error, or something else?

DennisDenuto replied:
not an error. it won't match (it will return an Anonymous) and will try the next keychain (iaas, cli, config.json etc)

cppforlife commented on pkg/imgpkg/registry/keychain_test.go:
```diff
@@ -169,6 +170,58 @@ func TestAuthProvidedViaEnvVars(t *testing.T) {
+		"localhost:9999/v1/",
+		"localhost:9999/v2/",
+		"*:9999/v2/",
+		"local*:9999/v2/",
```
can we throw in a test for multi domain globs: *.*.foo.com?

DennisDenuto replied:
yep done!

cppforlife commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -49,7 +50,12 @@ func (k *envKeychain) Resolve(target regauthn.Resource) (regauthn.Authenticator,
 	}
 
 	for _, info := range infos {
```
i think we need to consider how should ordering of infos works out here e.g.

foo.com/image

vs

foo.com

i think we would want to be matching more specific cases. do we have some precedent in credential provider hows thats done?

DennisDenuto replied:
that makes sense... i'll add some ordering logic where the more specific match should take precedence.

DennisDenuto replied:
took a very similar approach to what `BasicDockerKeyring` does in terms of handling multiple urls.

https://github.com/vmware-tanzu/carvel-imgpkg/blob/3a779ca0886ab989a22c6928a1eb58cd36dc0bbd/vendor/github.com/vdemeester/k8s-pkg-credentialprovider/keyring.go#L42

DennisDenuto committed:
feat: Allow specifying a scheme to the hostname

- Also considers url path containing /v1/ or /v2/ to be the equivalent
to the hostname

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: Allow matching HOSTNAME by globs

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: Allow matching HOSTNAME by repository

- Add ordering when deciding which hostname matches the target image
-- done via the various env hostnames being sorted, resulting in the more specific
paths matching first
- renamed info.Hostname -> info.URL since it now stores a URL and not
*always* just a hostname

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
@cppforlife as discussed, merging into develop since tests went green.
